### PR TITLE
1.x: optimize collect, reduce and takeLast(1)

### PR DESCRIPTION
**Pull request by akarnokd**
This PR reduces the overhead with `collect`, `reduce` and `takeLast` and introduces the `DeferredScalarSubscriber` to support similar unbounded-in single-out scenarios.

Benchmark comparison (i7 4770K, Windows 7 x64, Java 8u92)

image

Relative to each other: (each practically returning the very last value)

image

### CODE REVIEW

File: src/main/java/rx/internal/operators/DeferredScalarSubscriber.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
+/**
+ * Copyright 2015 Netflix, Inc.
```
Comment (JakeWharton) [inline src/main/java/rx/internal/operators/DeferredScalarSubscriber.java]: 2016? The class this came from had 2016 as well

**Comment by akarnokd** [comment]
Updated header year, @JakeWharton .

**Comment by JakeWharton** [comment]
lgtm :+1:

**Pull request merged**